src/eager_unix.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+use resources::{EagerAccept, EagerRead, EagerWrite, Resource};
+use tokio_util;
+use tokio_write;
+
+use futures::future::{self, Either};
+use std;
+use std::io::{ErrorKind, Read, Write};
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd};
+use tokio;
+use tokio::net::{TcpListener, TcpStream};
+use tokio_io;
+
+pub fn tcp_read<T: AsMut<[u8]>>(
+  tcp_stream: &TcpStream,
+  resource: Resource,
+  mut buf: T,
+) -> EagerRead<Resource, T> {
+  // Unforunately we can't just call read() on tokio::net::TcpStream
+  let fd = (*tcp_stream).as_raw_fd();
+  let mut std_tcp_stream = unsafe { std::net::TcpStream::from_raw_fd(fd) };
+  let read_result = std_tcp_stream.read(buf.as_mut());
+  // std_tcp_stream will close when it gets dropped. Thus...
+  let _ = std_tcp_stream.into_raw_fd();
+  match read_result {
+    Ok(nread) => Either::B(future::ok((resource, buf, nread))),
+    Err(err) => {
+      if err.kind() == ErrorKind::WouldBlock {
+        Either::A(tokio_io::io::read(resource, buf))
```

Comment:
@ry
One of the downsides of calling read(resource) rather than read(tcp_stream) here is that the resource has to be looked up again in the resource table, including the mutex action that comes with it. 
Would be nicer to call read on the tcp_stream directly.

Reply:
Yes would be nicer- I suspect it leads to type insanity tho.

Reply:
The problem is that tokio_io wants to store the object that has read from with the result (buf and nread). But whether that 'object' is a Resource or a TcpStream matters - they're both very different objects.

It seems that we're not using the Resource reference though, maybe we can just throw it away?

Reply:
It's safe to drop Resource - it's just a wrapper around an integer.

Reply:
I wasn't able to figure it out, you're right, type insanity...